### DISCUSSION TITLE
Newbie trying to upgrade from 2.10.5 and stuck at import error due to 'soft_unicode' ?

### DISCUSSION THREAD

**Discussion by 1920x1200** [netbox-community/netbox#9717] [Q&A] [answered]
Hi everyone, I just took over a NetBox install from a previous employee and it hasn't been upgraded for over a year. It's currently running 2.10.5 and I am trying to get it upgraded to the latest. The server is Ubuntu 20.04.2 LTS and has Python 3.8.10 and PostgreSQL 12.9. I'm following the process in [this video](https://www.youtube.com/watch?v=2tlZcjtDxBI) and I am stuck at the error:

`ImportError: cannot import name 'soft_unicode' from 'markupsafe' (/opt/netbox/venv/lib/python3.8/site-packages/markupsafe/__init__.py)`

Attached is the full log. [netbox upgrade attempt.txt](https://github.com/netbox-community/netbox/files/9105285/netbox.upgrade.attempt.txt)

I've seen some other people run into this same error and it seems like it has to do with a incompatible version of MarkupSafe but when I try to downgrade it, I run into the same problem. 

I first tried to upgrade from 2.10.5 to 2.10.6 and also tried going straight to 2.11.12 and ran into the same issue. I read that some folks just give up and create a new instance of 2.11.12, import their database, and then do the upgrades but I'd like to avoid that if possible. 

I'm a network admin and don't have much Linux experience so I'm hoping someone can help me out or point me in the right direction. 

Thanks,

Scott

**Comment by candlerb** [comment, marked as answer]
> Attached is the full log

Thanks for providing the full log, as it makes it clear what's happening.

You're running some script (that you found on the Internet?) that outputs some shell commands.  You're then blinding pasting that in, all at once, and ignoring all the errors it returns.

I suggest you try pasting it one line at a time, looking at the errors returned from that command, and fixing them before proceeding to the next one.  For example, the first command:

```
wget https://github.com/netbox-community/n...2.11.12.tar.gz
```

generates the following response:

```
--2022-07-13 19:09:51--  https://github.com/netbox-community/n...2.11.12.tar.gz
Resolving github.com (github.com)... 140.82.113.4
Connecting to github.com (github.com)|140.82.113.4|:443... connected.
HTTP request sent, awaiting response...
404 Not Found
2022-07-13 19:09:51 ERROR 404: Not Found.
```

There's a strong clue as to what is happening. Why do you think it might be?

Once you've understood that, it should be pretty clear why the next command fails:

```
sudo tar -xzf v2.11.12.tar.gz -C /opt
```

Response:

```
tar (child): v2.11.12.tar.gz: Cannot open: No such file or directory
tar (child): Error is not recoverable: exiting now
```

If you don't understand any of the above, then I think you'd be best served by a local friendly sysadmin who can help you step through it line by line.

> I read that some folks just give up and create a new instance of 2.11.12, import their database, and then do the upgrades but I'd like to avoid that if possible.

It's a safe and robust approach. It's certainly less risky than an in-place upgrade, which can leave your system in a broken state if something goes wrong (although if you have VM snapshots that you can roll back to, that mitigates the risk)

**Reply by DanSheps** [reply]
If you decide to do a new instance, pull from the repo instead of using tarballs.

It makes performing upgrades much easier.

**Reply by 1920x1200** [reply]
Hi @candlerb, thanks for your reply. That all makes sense now and the issue was the link in that YouTuber's script was truncated and that's why it kept failing. I compared it to the NetBox documentation and corrected that, also verified the other steps so now I have a better idea of what the script is actually doing. I then went line by line but still got the error about the 'soft_unicode' from 'markupsafe' but I was able to correct that as well. I edited the local_requirements.txt file and added "MarkupSafe==2.0.1" and then the upgrade script seemed to run successfully! I run the command to restart it and then go to the IP address of my NetBox and I'm able to log in, but it's still running the old version. I've tested again upgrading from 2.10.5 to 2.10.6 and to 2.11.12 but it's not clear how the upgrade script knows which version it should go to or where it is even pulling the files from, so I think it's just re-installing the current version. Is there something I'm missing?

**Reply by candlerb** [reply]
> I run the command to restart it and then go to the IP address of my NetBox and I'm able to log in, but it's still running the old version

Check the output of:

```
ls -l /opt/netbox
```

to see where the symlink points for the "current version"; it should show `-> /opt/netbox-x.x.x`

If your `netbox.service` is the [standard one](https://github.com/netbox-community/netbox/blob/develop/contrib/netbox.service), then it should be pointing to `/opt/netbox/venv/bin/gunicorn`, so the symlink determines which version is run.

> it's not clear how the upgrade script knows which version it should go to or where it is even pulling the files from

It's using the files in the directory where the upgrade script is run. It knows which database migrations have not yet been applied by looking at state in a special table in the database itself.

**Reply by 1920x1200** [reply]
Hmm, I don't seem to have anything for "current version" ?
```
root@netbox:~# ls -l /opt/netbox
total 32284
-rw-r--r--  1 root root      117 Feb 25  2021 CHANGELOG.md
-rw-r--r--  1 root root     9278 Feb 25  2021 CONTRIBUTING.md
-rw-r--r--  1 root root    10174 Feb 25  2021 LICENSE.txt
-rw-r--r--  1 root root       72 Feb 25  2021 NOTICE
-rw-r--r--  1 root root     2480 Feb 25  2021 README.md
-rw-r--r--  1 root root     2438 Feb 25  2021 base_requirements.txt
drwxr-xr-x  8 root root     4096 Mar  6  2021 ciscodnacnetbox
drwxr-xr-x  2 root root     4096 Feb 25  2021 contrib
drwxr-xr-x 13 root root     4096 Feb 25  2021 docs
-rw-r--r--  1 root root 32941154 Mar  6  2021 dumpfile
-rw-r--r--  1 root root      486 Feb 25  2021 gunicorn.py
-rwxrwxrwx  1 root root       58 Jul 14 19:04 local_requirements.txt
-rw-r--r--  1 root root    16084 Feb 27  2021 master.zip
-rw-r--r--  1 root root     4129 Feb 25  2021 mkdocs.yml
drwxr-xr-x 18 root root     4096 Mar  6  2021 netbox
lrwxrwxrwx  1 root root       19 Jul 14 18:53 netbox-2.10.6 -> /opt/netbox-2.10.6/
lrwxrwxrwx  1 root root       19 Jul 13 15:59 netbox-2.11.0 -> /opt/netbox-2.11.0/
lrwxrwxrwx  1 root root       20 Jul 14 18:56 netbox-2.11.12 -> /opt/netbox-2.11.12/
drwxr-xr-x  4 root root     4096 Feb 27  2021 netbox-scanner
-rw-r--r--  1 root root      654 Mar  9  2021 netbox-scanner.conf
-rw-r--r--  1 root root      470 Feb 25  2021 requirements.txt
drwxr-xr-x  3 root root     4096 Feb 25  2021 scripts
-rwxr-xr-x  1 root root     3616 Mar 21 19:09 upgrade.sh
drwxr-xr-x  6 root root     4096 Jul 14 19:04 venv
root@netbox:~# 
```

**Reply by candlerb** [reply]
Do you have a hidden directory `/opt/netbox/.git`?

```
ls -ld /opt/netbox/.git
```

If so, you're following the wrong upgrade instructions.  If Netbox was originally installed using git, then you must use the [git upgrade instructions](https://docs.netbox.dev/en/stable/installation/upgrading/#option-b-clone-the-git-repository).  The method you've been using is only if Netbox was [installed from a release tarball](https://docs.netbox.dev/en/stable/installation/upgrading/#option-a-download-a-release).

**Reply by 1920x1200** [reply]
!Solved Turns out this instance of NetBox was built via Git and I was following the wrong upgrade process all along. I was able to successfully update one minor version and now I understand the process to do the rest. Thanks for all your help @candlerb, I really appreciate you taking the time to help me with this as I learn these new skills!!!
